### ISSUE TITLE
Adding a shallow detection head to the model greatly increases the model training time. Why is this?

### ISSUE THREAD

**Issue by GZPLHJ** [ultralytics/ultralytics#6966]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

When I use Yolov8x.yaml to train the model, the image input size is 1024*1024. One training round takes 15 minutes.
When I made changes to yolov8x.yaml and added an additional detection head, the image input size remained unchanged, and the training time increased to 3 hours.
When I select the s model size for model width and depth, my parameter amount and calculation amount are much lower than Yolov8x.yaml, and the training time is also 50 minutes per round.
My graphics card choice is 4090.
may I know what is the reason?

**Comment by GZPLHJ** [comment]
When using yolov8x.yaml (3 detection heads):
**********************************************************************************************************************************
# Ultralytics YOLO 🚀, GPL-3.0 license

# Parameters
nc: 80  # number of classes
depth_multiple: 1.00  # scales module repeats
width_multiple: 1.25  # scales convolution channels
D:\GuoZhoupeng\object_detect\yolov8\ultralytics\models\v8\yolov8x:
  yaml:
# YOLOv8.0x backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]]  # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]]  # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, Conv, [256, 3, 2]]  # 3-P3/8
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]]  # 5-P4/16
  - [-1, 6, C2f, [512, True]]
  - [-1, 1, Conv, [512, 3, 2]]  # 7-P5/32
  - [-1, 3, C2f, [512, True]]
  - [-1, 1, SPPF, [512, 5]]  # 9

# YOLOv8.0x head
head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C2f, [512]]  # 12

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 4], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C2f, [256]]  # 15 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C2f, [512]]  # 18 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [512]]  # 21 (P5/32-large)

  - [[15, 18, 21], 1, Detect, [nc]]  # Detect(P3, P4, P5)
![image](https://github.com/ultralytics/ultralytics/assets/57591529/62222870-b089-4e99-8b13-41acb1b14882)
The yolov8x model parameters here are very large。

![image](https://github.com/ultralytics/ultralytics/assets/57591529/bc08e68e-4986-46a8-abcc-f5002ae5ee0b)
**********************************************************************************************************************************
When using yolov8x-4.yaml (4 detection heads):
**********************************************************************************************************************************
# Ultralytics YOLO 🚀, GPL-3.0 license

# Parameters
nc: 3  # number of classes
depth_multiple: 0.33  # scales module repeats
width_multiple: 0.50  # scales convolution channels
D:\GuoZhoupeng\object_detect\yolov8\ultralytics\models\v8\yolov8x:
  yaml:
# YOLOv8.0x backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]]  # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]]  # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, Conv, [256, 3, 2]]  # 3-P3/8
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]]  # 5-P4/16
  - [-1, 6, C2f, [512, True]]
  - [-1, 1, Conv, [512, 3, 2]]  # 7-P5/32
  - [-1, 3, C2f, [512, True]]
  - [-1, 1, SPPF, [512, 5]]  # 9

# YOLOv8.0x head
head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 6], 1, Concat, [1]]  # cat backbone P4
  - [-1, 3, C2f, [512]]  # 12

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 4], 1, Concat, [1]]  # cat backbone P3
  - [-1, 3, C2f, [256]]  # 15

  - [ -1, 1, nn.Upsample, [ None, 2, 'nearest' ] ]
  - [ [ -1, 2 ], 1, Concat, [ 1 ] ]
  - [ -1, 3, C2f, [ 128 ] ]   # 18



  - [-1, 1, Conv, [128, 3, 2]]
  - [[-1, 15], 1, Concat, [1]]
  - [-1, 3, C2f, [256]]  # 21 (P3/16-medium)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  # cat head P4
  - [-1, 3, C2f, [512]]  # 24 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [512]]  # 27 (P5/32-large)



  - [[18, 21, 24, 27], 1, Detect, [nc]]  # Detect(P3, P4, P5)
 
Here we have greatly reduced the width and depth of the model。
![image](https://github.com/ultralytics/ultralytics/assets/57591529/ddd6e3d2-9bbe-4ce1-9258-5df1ec20952a)

![image](https://github.com/ultralytics/ultralytics/assets/57591529/2a62d74f-5e05-42c6-b743-6a6fe625b779)


Even though the amount of parameter calculation is very small, the training speed is still very slow. Why is this?

**Comment by glenn-jocher** [comment]
@GZPLHJ adding additional heads to a YOLOv8 model increases the complexity of the network, especially computational complexity during training. Each additional detection head requires the model to perform more convolution operations and potentially additional feature concatenations, leading to more parameters to learn and gradients to compute. Even though GPU memory is often the limiting factor, computational complexity can significantly impact the training time as well.

When you add an extra head without adjusting depth and width multiples accordingly, you're increasing the workload for the entire network. The 4090 GPU is powerful, but adding more heads increases GPU memory usage, which might lead to longer training times or out-of-memory errors. Always ensure your network's complexity matches the computational resources you have available to maintain optimal training times. If you're experiencing significantly longer training times or memory issues, consider optimizing the depth and width of your model or using gradient checkpointing to reduce memory usage.

**Comment by GZPLHJ** [comment]
> @GZPLHJ adding additional heads to a YOLOv8 model increases the complexity of the network, especially computational complexity during training. Each additional detection head requires the model to perform more convolution operations and potentially additional feature concatenations, leading to more parameters to learn and gradients to compute. Even though GPU memory is often the limiting factor, computational complexity can significantly impact the training time as well.
> 
> When you add an extra head without adjusting depth and width multiples accordingly, you're increasing the workload for the entire network. The 4090 GPU is powerful, but adding more heads increases GPU memory usage, which might lead to longer training times or out-of-memory errors. Always ensure your network's complexity matches the computational resources you have available to maintain optimal training times. If you're experiencing significantly longer training times or memory issues, consider optimizing the depth and width of your model or using gradient checkpointing to reduce memory usage.

You're right, adding detection heads does increase complexity, but you didn't read the entirety of my question。
![image](https://github.com/ultralytics/ultralytics/assets/57591529/f01b6072-cbd4-4145-9085-feeb8d532cdf)
The amount of model parameters and calculations in this picture are much smaller than that of Yolov8x, because I use the width and depth of Yolov8s here, and I use 4 detection heads here：
depth_multiple: 0.33 # scales module repeats
width_multiple: 0.50 # scales convolution channels
It took me as long as 50 minutes to train this model.
![image](https://github.com/ultralytics/ultralytics/assets/57591529/0627a1dd-5af3-4542-918c-98153040b85b)
And it finally shows that I don’t have enough GPU memory? ? ? ? ?
But in comparison, I directly used the Yolov8x model to train, and it only took 14 minutes for one round.
![image](https://github.com/ultralytics/ultralytics/assets/57591529/d82d204b-0c36-47a8-9740-6ffa5a5a8c0c)
I have found the same problem with other users of other websites. I hope you can answer it. Thank you very much for your efforts! !

**Comment by glenn-jocher** [comment]
@GZPLHJ increasing the number of detection heads on your model, even with reduced width and depth multipliers, involves a trade-off between model size and training time. Even a relatively small increase in parameters and computational graph complexity can lead to disproportionate increases in memory requirements due to the way memory is allocated and managed during training, which includes not just the model parameters but also activations, gradients, and optimizer state.

The training duration discrepancy you are observing may be influenced by several factors. Firstly, the overall efficiency of the model with more heads can be lower despite fewer parameters, due to less optimal use of the GPU's parallel processing capabilities. Additionally, certain operations can have overheads that don't scale linearly with the number of parameters.

The memory error suggests that the combined effect of the added detection head and the existing architecture's resource demands exceeds what your GPU can handle, possibly due to memory fragmentation or a temporary spike in memory usage that exceeds the capacity.

I would suggest revisiting the model architecture and considering whether you can simplify some parts of the model or employ techniques such as mixed-precision training to reduce memory usage. It's also beneficial to monitor GPU memory usage during training to identify precisely when and where the out-of-memory issue occurs, which might offer further insights into possible optimizations.

**Comment by GZPLHJ** [comment]
> @GZPLHJ increasing the number of detection heads on your model, even with reduced width and depth multipliers, involves a trade-off between model size and training time. Even a relatively small increase in parameters and computational graph complexity can lead to disproportionate increases in memory requirements due to the way memory is allocated and managed during training, which includes not just the model parameters but also activations, gradients, and optimizer state.
> 
> The training duration discrepancy you are observing may be influenced by several factors. Firstly, the overall efficiency of the model with more heads can be lower despite fewer parameters, due to less optimal use of the GPU's parallel processing capabilities. Additionally, certain operations can have overheads that don't scale linearly with the number of parameters.
> 
> The memory error suggests that the combined effect of the added detection head and the existing architecture's resource demands exceeds what your GPU can handle, possibly due to memory fragmentation or a temporary spike in memory usage that exceeds the capacity.
> 
> I would suggest revisiting the model architecture and considering whether you can simplify some parts of the model or employ techniques such as mixed-precision training to reduce memory usage. It's also beneficial to monitor GPU memory usage during training to identify precisely when and where the out-of-memory issue occurs, which might offer further insights into possible optimizations.
Thank you very much for your reply. I probably know where to start to check this problem. I will continue to observe and study this issue in the future. I will explain it in the reply area when I find a solution.

**Comment by GZPLHJ** [comment]
Thank you very much for your reply. I probably know where to start to check this problem. I will continue to observe and study this issue in the future. I will explain it in the reply area when I find a solution.

**Comment by glenn-jocher** [comment]
@GZPLHJ You're welcome! I'm glad to hear you have a starting point to investigate the issue further. Should you discover the root cause or need additional assistance in the future, don't hesitate to reach out. Happy coding, and we look forward to any solutions or insights you might share. 🚀
